reference/src/main/java/oshi/driver/ComponentDriver.java

```diff
@@ -0,0 +1,169 @@
+package oshi.driver;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import oshi.api.NoSuchQueryMethodException;
+
+/**
+ * A {@link ComponentDriver} is the root of the driver hierarchy. Containers can
+ * call {@link #query(Enum)} or {@link #queryAll()} to have this class update
+ * their state. There is a one-to-one correspondence between
+ * {@link ComponentDriver}s and container objects.<br>
+ * <br>
+ * This class automatically registers all query methods in its subclasses. Query
+ * methods in {@link ExtensionDriver}s can also be registered with
+ * {@link #register(ExtensionDriver)}.
+ * 
+ * @author cilki
+ * @since 5.0.0
+ */
+public class ComponentDriver {
+
+	private static final class QueryHandle implements Comparable<QueryHandle> {
+		public MethodHandle handle;
+		public Fallback fallback;
+		public Class<?> cls;
+
+		public QueryHandle(MethodHandle handle, Fallback fallback, Class<?> cls) {
+			this.handle = handle;
+			this.fallback = fallback;
+			this.cls = cls;
+		}
+
+		@Override
+		public int compareTo(QueryHandle o) {
+			// TODO implement query ordering with fallbacks
+			return 0;
+		}
+	}
+
+	/**
+	 * An association between attribute enums and lists of query handles in
+	 * descending priority order. A particular query handle can exist in multiple
+	 * lists, but not more than once in each list.
+	 */
+	private Map<Enum<?>, List<QueryHandle>> handles;
+
+	/**
+	 * A list of extensions in case they are needed in the future.
+	 */
+	private List<ExtensionDriver> extensions;
+
+	public ComponentDriver() {
+		this.handles = new HashMap<>();
+		this.extensions = new LinkedList<>();
```

Comment:
You should [generally almost always use an `ArrayList<>`](https://stackoverflow.com/questions/322715/when-to-use-linkedlist-over-arraylist-in-java) unless you have a very specific use case for which LinkedLists is better (e.g., a high volume of adds / removes at only the head or tail).

Reply:
I went with `LinkedList` to avoid having to call `trimToSize` later since that list will never change after initialization. I'm actually not sure if the list is even necessary, but it's there just in case. I'll see if I can use `Collections.unmodifiableList` which would be even better.